sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.ClientModel.Options;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// TODO.
+/// </summary>
+public class ClientLoggingPolicy : PipelinePolicy
+{
+    private const double RequestTooLongTime = 3.0; // sec
+    private const int DefaultLoggedContentSizeLimit = 4 * 1024;
+    private static ClientModelEventSource? s_eventSource;
+    private const string DefaultEventSourceName = "System.ClientModel";
+
+    private readonly bool _logContent;
+    private readonly int _maxLength;
+    private readonly string? _assemblyName;
+    private readonly string? _clientRequestIdHeaderName;
+    private readonly bool _isLoggingEnabled;
+
+    internal static ClientModelEventSource Singleton { get => s_eventSource ??= ClientModelEventSource.Create(DefaultEventSourceName, Array.Empty<string>()); } //TODO add traits
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="eventSourceName"></param>
+    /// <param name="eventSourceTraits"></param>
+    /// <param name="options"></param>
+    protected ClientLoggingPolicy(string? eventSourceName, string[]? eventSourceTraits = default, DiagnosticsOptions? options = default) : this(options)
+    {
+        if (s_eventSource != null && !string.IsNullOrEmpty(eventSourceName) && eventSourceName != s_eventSource.Name)
+        {
+            throw new ArgumentException("Cannot use multiple event source names in the same application.");
+        }
+
+        if (s_eventSource == null)
+        {
+            s_eventSource = ClientModelEventSource.Create(eventSourceName ?? DefaultEventSourceName, eventSourceTraits);
```

Comment:
When/if we move forward with the static dictionary of singletons, it would be good to mock up an "Azure.Core 2.0" implementation -- i.e. a wrapper logging policy that can customize the name and traits and maybe also a mini-app that validates the scenario where we have multiple clients logging to both "Azure-Core" and "System-ClientModel" event sources.  I'm interested to see how well the traits part works, if it looks clunky, etc.

Reply:
Added some tests that do a similar thing in ClientLoggingPolicyTests